dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexReader.cpp

```diff
@@ -0,0 +1,156 @@
+#include <Storages/DeltaMerge/Index/VectorIndexCache.h>
+#include <Storages/DeltaMerge/Index/VectorSearchPerf.h>
+
+
+namespace DB::DM
+{
+
+void ColumnFileTinyVectorIndexReader::read(
+    MutableColumnPtr & vec_column,
+    const std::span<const VectorIndexViewer::SearchResult> & selected_rows,
+    size_t start_offset,
+    size_t column_size)
+{
+    RUNTIME_CHECK(loaded);
+
+    Stopwatch watch;
+    vec_column->reserve(column_size);
+    std::vector<Float32> value;
+    size_t current_rowid = start_offset;
+    for (const auto & [rowid, _] : selected_rows)
+    {
+        vec_index->get(rowid, value);
+        if (rowid > current_rowid)
+        {
+            UInt32 nulls = rowid - current_rowid;
+            // Insert [] if column is Not Null, or NULL if column is Nullable
+            vec_column->insertManyDefaults(nulls);
+        }
+        vec_column->insertData(reinterpret_cast<const char *>(value.data()), value.size() * sizeof(Float32));
+        current_rowid = rowid + 1;
+    }
+    if (current_rowid < start_offset + column_size)
+    {
+        UInt32 nulls = start_offset + column_size - current_rowid;
+        vec_column->insertManyDefaults(nulls);
+    }
+
+    perf_stat.returned_rows = selected_rows.size();
+    perf_stat.read_vec_column_seconds = watch.elapsedSeconds();
+}
+
+std::vector<VectorIndexViewer::SearchResult> ColumnFileTinyVectorIndexReader::load()
+{
+    if (loaded)
+        return {};
+
+    Stopwatch watch;
+
+    loadVectorIndex();
+    auto search_results = loadVectorSearchResult();
+
+    perf_stat.load_vec_index_and_results_seconds = watch.elapsedSeconds();
+
+    loaded = true;
+    return search_results;
+}
+
+void ColumnFileTinyVectorIndexReader::loadVectorIndex()
+{
+    const auto & index_infos = tiny_file.index_infos;
+    if (!index_infos || index_infos->empty())
+        return;
+    auto index_id = ann_query_info->index_id();
+    const auto index_info_iter
+        = std::find_if(index_infos->cbegin(), index_infos->cend(), [index_id](const auto & info) {
+              if (!info.vector_index)
+                  return false;
+              return info.vector_index->index_id() == index_id;
+          });
+    if (index_info_iter == index_infos->cend())
+        return;
+    auto vector_index = index_info_iter->vector_index;
+    if (!vector_index)
+        return;
+    auto index_page_id = index_info_iter->index_page_id;
+    auto load_from_page_storage = [&]() {
+        perf_stat.load_from_cache = false;
+        std::vector<size_t> index_fields = {0};
+        auto index_page = data_provider->readTinyData(index_page_id, index_fields);
+        ReadBufferFromOwnString read_buf(index_page.data);
+        CompressedReadBuffer compressed(read_buf);
+        return VectorIndexViewer::load(*vector_index, compressed);
+    };
+    if (vec_index_cache)
+    {
+        const auto key = fmt::format("{}{}", VectorIndexCache::COLUMNFILETINY_INDEX_NAME_PREFIX, index_page_id);
```

Comment:
vec_index_cache is global, maybe a single index_page_id is not enough in disaggregated mode?

Reply:
Add `keyspace_id`